xmtp/README.md

```diff
@@ -0,0 +1,104 @@
+  - Even simpler, we could have singleton threads for processMessages() and processPayloads() that simply run on an interval.
+
+### Creating a conversation
+
+```
+createConversation():
+    Deterministically derive convo_id from peer_address
+    If conversation with convo_id doesn't already exist in DB, insert it with conversation.state = UNINITIALIZED
+    return conversation
+```
+
+### Sending a message in a conversation
+
+```
+sendMessage():
+    Insert message into DB with message.state = UNINITIALIZED and message.convo_id set
+    processMessages() // Could be kicked off asynchronously or synchronously
+    return success
+
+processMessages():
+    For each message in UNINITIALIZED state, processing in order of timestamp:
+        Fetch the conversation's users from the DB (including self)
+        For each user:
+            If user.last_refreshed is uninitialized or more than THRESHOLD ago:
+                refreshUserInstallations()    // Could be kicked off asynchronously or synchronously
+                return  // refreshUserInstallations() will call back into processMessages() when ready
+        Fetch the installations of both users from the DB
+        For each installation:
+            // Build the plaintext payload
+            If conversation.state == UNINITIALIZED:
+                Construct the payload as an invite with the message attached
+            Else if conversation.state == INVITED:
+                Construct the payload as a message
+            // Encrypt the payload
+            If installation.state == UNINITIALIZED:
+                Create an outbound session (hold it in memory)
+            Use the existing session to encrypt a payload containing an invite with the message attached (hold it in memory)
+        In a single transaction:
+            Push all encrypted payloads to outbound_payloads table with outbound_payload.state = UNINITIALIZED
+            Commit all updated session states to the DB with installation.state = SESSION_CREATED
+            Set message.state = LOCALLY_COMMITTED
+            Set conversation.state = INVITED
+            Delete all data from memory
+    processPayloads()   // Could be kicked off asynchronously or synchronously
+    return
+
+refreshUserInstallations(user):
+    Fetch installations/contact bundles for the user from the network
+    Fetch installations/contact bundles for the user from the DB
+    In a single transaction:
+        For each installation from the DB:
+            If it doesn't exist in the network contact bundles or is expired or revoked, delete it from the DB
+        For each installation from the network:
+            If it doesn't exist in the DB, insert it with installation.state = UNINITIALIZED
+        Set user.last_refreshed to NOW
+    processMessages();  // Could be kicked off asynchronously or synchronously
```

Comment:
In this function the outer loop is over `UNINITIALIZED` messages, which in this case there are not guaranteed to be any -- Which would mean the initialization of installations would not be processed until a message is sent.

Is this intentional?